ftw/statusmap/utils.py

```diff
@@ -48,3 +51,120 @@ def getInfos(context, cat, wf_tool):
     items = getBaseInfo(path, brains)
     items = getTransitionsForItem(wf_tool, brains, items)
     return items
+
+
+class StatusMapFolderTree(object):
+
+    def __init__(self, context):
+        self.context = context
+        self.wf_tool = api.portal.get_tool('portal_workflow')
+        self.catalog = api.portal.get_tool('portal_catalog')
+        self.portal_properties = api.portal.get_tool('portal_properties')
+
+    def __call__(self):
+        return self.generate_tree()
+
+    def generate_tree(self):
+        """Generates a nested tree.
+        - The root of the tree will be the context
+        - The sorting is the plonenavigation sorting
+
+        [{
+        'url': '/foo'
+        'nodes': [{
+            'url': '/foo/bar'
+            }]
+        },
+        'url': '/dummy'
+        }]
+        """
+        nodes = map(self._node, self._brains())
+        return self._make_tree_by_url(nodes)
+
+    def _node(self, brain):
+        """Converts a brain into a statusmap node
+        """
+        return dict(
+            url=get_view_url(brain)[1],
+            review_state=brain.review_state or '',
+            type=brain.portal_type,
+            icon=brain.getIcon,
+            title=brain.pretty_title_or_id(),
+            uid=brain.UID,
+            transitions=self._get_transitions_of_brain(brain),
+        )
+
+    def _action_item(self, action):
+        """Converts a workflow action object into
+        an item
+        """
+        return dict(
+            id=action['id'],
+            title=action['title'],
+            new_review_state=action.get('transition').new_state_id,
+        )
+
+    def _get_transitions_of_brain(self, brain):
+        """Returns a dict of all possible workflowactions of the brains
+        object
+
+        TODO: This function makes it slow. We need the object.
```

Comment:
I don't think it is possible to optimize that, since we would have to reliable guess the result of the transition guards without having the object, which is not possible in a generalized package as ftw.statusmap is :unamused: